facade-charm/src/charm.py

```diff
@@ -0,0 +1,192 @@
+        e.set_results({"updated": updated})
+
+    def _on_clear_action(self, e: ops.ActionEvent):
+        updated = []
+        if endpoint := e.params.get("endpoint"):
+            e.log(f"clearing endpoint {endpoint}")
+            if relations := self.model.relations[endpoint]:
+                updated = self._update(*relations, clear=True)
+            else:
+                e.log(f"no bindings on {endpoint}")
+        else:
+            e.log(f"clearing all endpoints")
+            updated = self._update(clear=True)
+        e.set_results({"cleared": updated})
+
+    def _on_update_relation(self, e: ops.RelationEvent):
+        self._update(e.relation)
+
+    def _update(self, *relation: ops.Relation,
+                clear=False):
+        to_update = list(relation) if relation else list(chain(*self.model.relations.values()))
+        updated = []
+        for relation in to_update:
+            if self._update_relation(relation, clear=clear):
+                updated.append(relation.name)
+        return updated
+
+    def _update_relation(self, relation: ops.Relation,
+                         clear=False, replace=False):
+        changed = False
+
+        app_databag = relation.data[self.app]
+        unit_databag = relation.data[self.unit]
+        if replace or clear:
+            if app_databag.keys():
+                app_databag.clear()
+                changed = True
+            if unit_databag.keys():
+                unit_databag.clear()
+                changed = True
+
+        if not clear:
+            app_data, unit_data = self._load_mock(relation.name)
+            if app_data and dict(app_databag) != app_data:
+                app_databag.update(app_data)
+                changed = True
+            if unit_data and dict(unit_databag) != unit_data:
+                unit_databag.update(unit_data)
+                changed = True
+        return changed
+
+    def _load_mock(self, endpoint: str):
+        mocks_root = Path(__file__).parent.parent / 'mocks'
+
+        if endpoint.startswith("provide"):
+            pth = mocks_root / "provide" / (endpoint + ".yaml")
+        else:
+            pth = mocks_root / "provide" / (endpoint + ".yaml")
+
+        if not pth.exists():
+            logger.warning(f"mock not found for {endpoint} ({pth})")
+            pth.touch()
+            self._write_mock(endpoint, {}, {})
+            return {}, {}
+
+        yml = yaml.safe_load(pth.read_text())
+        app_data = yml.get("app_data", {})
+        unit_data = yml.get("unit_data", {})
+        return app_data, unit_data
+
+    def _write_mock(self, endpoint: str,
+                    app_data: Optional[dict] = None,
+                    unit_data: Optional[dict] = None):
+        mocks_root = Path(__file__).parent.parent / 'mocks'
+
+        if endpoint.startswith("provide"):
+            pth = mocks_root / "provide" / (endpoint + ".yaml")
+        else:
+            pth = mocks_root / "provide" / (endpoint + ".yaml")
```

Comment:
the two branches look the same, which would be odd

Reply:
very true, thanks!